gateway/mw_graphql_granular_access_test.go

```diff
@@ -110,3 +110,314 @@ func TestGraphQL_RestrictedTypes(t *testing.T) {
 		}...)
 	})
 }
+
+func TestGraphQL_AllowedTypes(t *testing.T) {
+	g := StartTest(nil)
+	defer g.Close()
+
+	api := g.Gw.BuildAndLoadAPI(func(spec *APISpec) {
+		spec.Proxy.ListenPath = "/"
+		spec.UseKeylessAccess = false
+		spec.GraphQL.Enabled = true
+	})[0]
+
+	_, directKey := g.CreateSession(func(s *user.SessionState) {
+		s.AccessRights = map[string]user.AccessDefinition{
+			api.APIID: {
+				APIID:       api.APIID,
+				APIName:     api.Name,
+				EnableAllow: true,
+				AllowedTypes: []graphql.Type{
+					{
+						Name:   "Country",
+						Fields: []string{"code"},
+					},
+					{
+						Name:   "Query",
+						Fields: []string{"countries"},
+					},
+				},
+			},
+		}
+	})
+
+	pID := g.CreatePolicy(func(p *user.Policy) {
+		p.AccessRights = map[string]user.AccessDefinition{
+			api.APIID: {
+				APIID:       api.APIID,
+				APIName:     api.Name,
+				EnableAllow: true,
+				AllowedTypes: []graphql.Type{
+					{
+						Name:   "Country",
+						Fields: []string{"name"},
+					},
+					{
+						Name:   "Query",
+						Fields: []string{"countries"},
+					},
+				},
+			},
+		}
+	})
+
+	_, policyAppliedKey := g.CreateSession(func(s *user.SessionState) {
+		s.ApplyPolicies = []string{pID}
+	})
+
+	t.Run("Direct key", func(t *testing.T) {
+		authHeaderWithDirectKey := map[string]string{
+			headers.Authorization: directKey,
+		}
+
+		allowedQuery := graphql.Request{
+			Query: "query Query { countries { code } }",
+		}
+
+		restrictedQuery := graphql.Request{
+			Query: "query Query { countries { name } }",
+		}
+
+		_, _ = g.Run(t, []test.TestCase{
+			{
+				Data:    restrictedQuery,
+				Headers: authHeaderWithDirectKey,
+				BodyMatchFunc: func(bytes []byte) bool {
+					return assert.Contains(t, string(bytes), `{"errors":[{"message":"field: name is restricted on type: Country"}]}`)
+				},
+				Code: http.StatusBadRequest,
+			},
+			{Data: allowedQuery, Headers: authHeaderWithDirectKey, Code: http.StatusOK},
+		}...)
+	})
+
+	t.Run("Policy applied key", func(t *testing.T) {
+		test.Flaky(t) // TODO: TT-5220
```

Comment:
TODO here?

Reply:
I mentioned that in the description. The previous test has that flag and I keep it. https://github.com/TykTechnologies/tyk/blob/master/gateway/mw_graphql_granular_access_test.go#L86

Reply:
Sorry!